File: 69-sqrtx/sqrtx.py
class Solution:
    def mySqrt(self, x: int) -> int:
        if x == 0:
            return 0
        cur_num = 2
        prev_num = 1
        for i in range (1, x // 2):
            if cur_num*cur_num > x:
                break
            cur_num += 1
            prev_num += 1

        return prev_num

File: 69-sqrtx/sqrtx.cpp
class Solution {
public:
    // basic idea is we keep decreasing the search space
    // mid = (left + right) / 2
    // if mid * mid > x implies, we need to search in left 
    // half, otherwise need to search in right half
    int mySqrt(int x) {
        int left = 0, right = x;
        int res = 0;
        while (left <= right) {
            long int mid = left + (right - left) / 2;
            if (mid * mid > x) {
                right = mid - 1;
            } else if (mid * mid < x) {
                left = mid + 1;
                res = mid;
            } else {
                return mid;
            }
        }
        return res;
    }
};